p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,302 @@
+		s.Reset()
+		log.Debugf("expectd HolePunch_CONNECT message, got %s", msg.GetType())
+		return
+	}
+	obsRemote := addrsFromBytes(msg.ObsAddrs)
+	rtt := time.Since(tstart)
+
+	// send a SYNC message and attempt a direct connect after half the RTT
+	msg.Reset()
+	msg.Type = holepunch_pb.HolePunch_SYNC.Enum()
+	if err := w.WriteMsg(msg); err != nil {
+		s.Reset()
+		log.Errorf("failed to send SYNC message for hole punching, err: %s", err)
+		return
+	}
+	defer s.Close()
+
+	synTime := time.Duration(rtt.Milliseconds()/2) * time.Millisecond
+
+	// wait for sync to reach the other peer and then punch a hole for it in our NAT
+	// by attempting a connect to it.
+	select {
+	case <-time.After(synTime):
+		pi := peer.AddrInfo{
+			ID:    rp,
+			Addrs: obsRemote,
+		}
+		hs.holePunchConnectWithBackoff(pi)
+
+	case <-hs.ctx.Done():
+		return
+	}
+}
+
+func (hs *HolePunchService) handleNewStream(s network.Stream) {
+	log.Infof("got hole punch request from peer %s", s.Conn().RemotePeer().Pretty())
+	_ = s.SetDeadline(time.Now().Add(holePunchTimeout))
+	rp := s.Conn().RemotePeer()
+	wr := protoio.NewDelimitedWriter(s)
+	rd := protoio.NewDelimitedReader(s, maxMsgSize)
+
+	// Read Connect message
+	msg := new(holepunch_pb.HolePunch)
+	if err := rd.ReadMsg(msg); err != nil {
+		s.Reset()
+		return
+	}
+	if msg.GetType() != holepunch_pb.HolePunch_CONNECT {
+		s.Reset()
+		return
+	}
+	obsDial := addrsFromBytes(msg.ObsAddrs)
+
+	// Write CONNECT message
+	msg.Reset()
+	msg.Type = holepunch_pb.HolePunch_CONNECT.Enum()
+	msg.ObsAddrs = addrsToBytes(hs.ids.OwnObservedAddrs())
+	if err := wr.WriteMsg(msg); err != nil {
+		s.Reset()
+		return
+	}
+
+	// Read SYNC message
+	msg.Reset()
+	if err := rd.ReadMsg(msg); err != nil {
+		s.Reset()
+		return
+	}
+	if msg.GetType() != holepunch_pb.HolePunch_SYNC {
+		s.Reset()
+		return
+	}
+	defer s.Close()
+
+	// Hole punch now by forcing a connect
+	pi := peer.AddrInfo{
+		ID:    rp,
+		Addrs: obsDial,
+	}
+	hs.holePunchConnectWithBackoff(pi)
+}
+
+func (hs *HolePunchService) holePunchConnectWithBackoff(pi peer.AddrInfo) {
+	forceDirectConnCtx := network.WithForceDirectDial(hs.ctx, "hole-punching")
+	dialCtx, cancel := context.WithTimeout(forceDirectConnCtx, dialTimeout)
+	defer cancel()
+	err := hs.host.Connect(dialCtx, pi)
+	if err == nil {
+		log.Infof("hole punch with peer %s successful, direct conns to peer are:", pi.ID.Pretty())
+		for _, c := range hs.host.Network().ConnsToPeer(pi.ID) {
+			if !isRelayAddress(c.RemoteMultiaddr()) {
+				log.Info(c)
+			}
+		}
+		return
+	}
+
+	// backoff and retry before giving up.
+	// this code will make the peer retry for (approximately ) a TOTAL of 10 seconds
+	// before giving up and declaring the hole punch a failure.
+	b := &backoff.Backoff{
+		Jitter: true,
+		Min:    1 * time.Second,
+		Max:    5 * time.Second,
+		Factor: 2,
+	}
+	for b.Attempt() < maxRetries {
+		time.Sleep(b.Duration())
+		err = hs.host.Connect(dialCtx, pi)
```

Comment:
We might need to make a new context here, as `WithTimeout` might convert internally to an absolute deadline.
Not sure, but let's be on the safe side.

Reply:
done.